src/write_gcm.c

```diff
@@ -597,26 +591,21 @@ static int wg_curl_get_or_post(char *response_buffer,
   return result;
 }
 
-static size_t wg_curl_write_callback(char *ptr, size_t size, size_t nmemb,
+static size_t wg_curl_write_callback(void *ptr, size_t size, size_t nmemb,
                                      void *userdata) {
-  wg_curl_write_ctx_t *ctx = userdata;
-  if (ctx->size == 0) {
+  size_t requested_bytes = size * nmemb;
+  wg_curl_write_ctx_t *ctx = (wg_curl_write_ctx_t *) userdata;
+
+  ctx->data = realloc(ctx->data, ctx->size + requested_bytes + 1);
```

Comment:
A sequence of reallocs is going to exhibit quadratic runtime, unless you employ the customary "doubling" strategy or similar. I'd recommend keeping track of both 'size' and 'capacity' in wg_curl_write_ctx_t, and then doubling capacity (and doing the appropriate realloc) when you run out of space.

Reply:
The example here: https://curl.haxx.se/libcurl/c/getinmemory.html shows the implementation as we have it. I was under the impression that the "size" parameter would resemble the content-length returned by the server. Do you have any links / references that suggest otherwise? Keep in mind I'm a C newbie, so I'm genuinely curious.

Reply:
A sequence of `realloc`s is only quadratic if each `realloc` copies. With reasonable `malloc` implementations, there will be fragmentation, and many `realloc`s will be able to grow the chunks in-place. Given the bucketing of the free list by size, it's likely to achieve the same effect as doubling, but without the hassle.

Reply:
I really wanted to disagree with you (and with the CURL authors cited by bmoyles) but this little C program I wrote proves Igor's point (alas). Being copied 212 times is many more times than a doubling allocator would copy, but it's not horrible either.

In my defense, I'm not aware that this is documented (i.e. promised) anywhere. So you're relying on implementation-defined behavior. This doesn't make me 100% comfortable, but it doesn't make me 100% uncomfortable either. I'm in a comfort grey zone.

```
#include <stdio.h>
#include <stdlib.h>

int main() {
  char * p = malloc(1);
  int numRequests = 0;
  int numCopies = 0;
  for (int s = 2; s < 10000000; ++s) {
    char * newp = realloc(p, s);
    if (newp != p) {
      ++numCopies;
    }
    ++numRequests;
    p = newp;
  }
  printf("I requested %d reallocs, and the buffer was copied %d times\n",
    numRequests, numCopies);
  free(p);
}
```

```I requested 9999998 reallocs, and the buffer was copied 212 times```

Reply:
Oh, and to @bmoyles0117 's actual point: if you're saying that---in practice---the WriteCallback is only called back once anyway, I suppose that's possible. Out here in the peanut gallery I only complain about abstractions, not practicalities. It certainly still needs to operate correctly if called back more than once. But if it's typically only called once anyway, that reduces the need to care much about the doubling stuff.

(on the third hand, the doubling stuff is like 6 lines of code...)

Reply:
OK, on the fourth hand I now see that the answer crucially depends on what else malloc is doing. This new program adds an intervening malloc in the loop, and it's doing a lot more copies. I'm going to claim that Igor's claim is not trustworthy, in that it depends too much on other allocations that are going on:

```
$ cat test.c
#include <stdio.h>
#include <stdlib.h>

int main() {
  char * p = malloc(1);
  int numRequests = 0;
  int numCopies = 0;
  for (int s = 2; s < 10000000; ++s) {
    malloc(1);  // stir things up with another malloc
    char * newp = realloc(p, s);
    if (newp != p) {
      ++numCopies;
      printf("copied when going to size %d\n", s);
    }
    ++numRequests;
    p = newp;
  }
  printf("I requested %d reallocs, and the buffer was copied %d times\n",
    numRequests, numCopies);
  free(p);
}
```

copied when going to size 25
copied when going to size 41
copied when going to size 57
copied when going to size 73
copied when going to size 89
copied when going to size 105
copied when going to size 121
copied when going to size 137
copied when going to size 153
copied when going to size 169
copied when going to size 185
copied when going to size 201
copied when going to size 217
copied when going to size 233
copied when going to size 249
copied when going to size 265
copied when going to size 281
copied when going to size 297
copied when going to size 313
copied when going to size 329
...